### PR TITLE
Feature: Add cases and function branch

### PR DESCRIPTION
**Pull request by KFoxder**
## Overview

- Attempts to address https://github.com/ocaml-sys/config.ml/issues/4 and https://github.com/ocaml-sys/config.ml/issues/5 by applying the config to expressions. 
- Added test cases for cases and function branches
- Depends on https://github.com/ocaml-sys/config.ml/pull/11

### TODO

- [ ] Update documentation to show uses in cases and feature branches

### CODE REVIEW

**Comment by KFoxder** [comment]
@leostera & @gpetiot I have a couple questions before I think its ready for review:

**Styling:**
What is the best way for a variable naming convention in OCaml. I did take a look at https://ocaml.org/docs/guidelines but didn't see anything useful. Particularly I am asking around shadowing variable names like `exp` in the below example. I am aware it is "correct" in that the scoping works but just not confident that it is named as best it could be. How would you name the variables there? 
```ocaml
let rec apply_config_on_expression (exp : expression) =
  let pexp_desc =
    match exp.pexp_desc with
    | Pexp_try (exp, cases) ->
        let exp = apply_config_on_expression exp in
        let cases = apply_config_on_cases cases in
        Pexp_try (exp, cases)
```

**Logic:**
1. I didn't try to address every single case for [`expression_desc`](https://github.com/ocaml-ppx/ppxlib/blob/455f21787ad52a32813a34538a4d493f8d3efc64/astlib/ast_501.ml#L303-L444). I only addressed the ones I needed to add the use cases described in the issues. I definitely can add them all and flesh it out,  but before I did wanted to take a pause and get your feedback of whether I should in this PR or not. 
    - If they aren't added it could lead to situations where someone would expect it to work with the following code and it doesn't:
    ```ocaml
    (* Won't work because I am not covering the `Pexp_setfield` *)
    a.prop <- (match x with | "0" -> 0 | "1" -> 1 [@cfg (target_ox = "macos")])
    ```
3.  The `apply_config_on_expression` function right now is just concerned with removing cases and not concerned with removing entire expressions. Therefore I am not returning an `Option` monad. I think in the future if the `@cfg` is applied to an expression and we want it to remove the entire expression we may want to do that but felt like overkill for now. Thoughts?

**Comment by KFoxder** [comment]
@leostera this can be reviewed when you get a chance. Not a priority or rush though.

File: config/cfg_ppx.ml [hunk 1]
```diff
@@ -107,6 +107,45 @@ let apply_config_on_types (tds : type_declaration list) =
...
+    | Pexp_function cases ->
+        let cases = apply_config_on_cases cases in
+        Pexp_function cases
+    | Pexp_let (rec_flag, vbs, exp) ->
+      let exp = apply_config_on_expression exp in
+      Pexp_let (rec_flag, vbs, exp)
+    | _ -> exp.pexp_desc
+  in
+  let { pexp_loc; pexp_loc_stack; pexp_attributes; _ } = exp in
+  { pexp_loc; pexp_loc_stack; pexp_attributes; pexp_desc }
```
Comment (leostera) [inline config/cfg_ppx.ml]: we can use this patch to update the record without redefining everything:

```suggestion
  { exp with pexp_desc }
```

File: config/cfg_ppx.ml [hunk 2]
```diff
@@ -107,6 +107,45 @@ let apply_config_on_types (tds : type_declaration list) =
...
+
+let apply_config_on_value_bindings (vbs : value_binding list) =
+  List.map
+    (fun vb ->
+      let { pvb_loc; pvb_attributes; pvb_pat; pvb_expr } = vb in
+      let _pvb_expr = apply_config_on_expression pvb_expr in
+      { pvb_loc; pvb_attributes; pvb_pat; pvb_expr = _pvb_expr })
```
Comment (leostera) [inline config/cfg_ppx.ml]: same as above:

```suggestion
      { vb with pvb_expr = apply_config_on_expression vb.pvb_expr })
```

File: config/cfg_ppx.ml [hunk 3]
```diff
@@ -107,6 +107,45 @@ let apply_config_on_types (tds : type_declaration list) =
       | _ -> Some td)
     tds
 
+let apply_config_on_cases (cases : cases) =
+  List.filter
+    (fun case ->
+      if should_keep case.pc_rhs.pexp_attributes = `keep then true else false)
```
Comment (leostera) [inline config/cfg_ppx.ml]: the comparison would be enough here:

```suggestion
      should_keep case.pc_rhs.pexp_attributes = `keep)
```

**Review by leostera — APPROVED** [review]
Looks good to me! :sparkles: Some small comments to make the record updates more readable, but the premise works. 

Thanks for the PR @KFoxder! 🙏🏼

**Comment by KFoxder** [comment]
@leostera Thanks for the PR review on a Sunday! Hope you are enjoying NYC!

**Comment by leostera** [comment]
Thanks for the PR @KFoxder! ✨

**Pull request merged**